core/platform/email/gae_email_services_test.py

```diff
@@ -38,8 +44,144 @@ def test_sending_email(self):
+        messages = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAIL)
+        self.assertEqual(1, len(messages))
+        self.assertEqual(messages[0].bcc, feconf.ADMIN_EMAIL_ADDRESS)
+
+    def test_that_email_reply_to_email_is_sent_when_reply_to_id_is_true(self):
+        reply_to = 'reply_to_address'
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_mail(
+                feconf.SYSTEM_EMAIL_ADDRESS, feconf.ADMIN_EMAIL_ADDRESS,
+                'subject', 'body', 'html', bcc_admin=False,
+                reply_to_id=reply_to)
+        messages = self.mail_stub.get_sent_messages(
+            to=feconf.ADMIN_EMAIL_ADDRESS)
+        valid_reply_to = gae_email_services.get_incoming_email_address(reply_to)
+        self.assertEqual(
+            messages[0].reply_to, valid_reply_to)
+
+    def test_that_email_not_sent_if_sender_address_is_malformed(self):
+        malformed_sender_email = ''
+        email_exception = (
+            self.assertRaisesRegexp(
+                ValueError, 'Malformed sender email address: %s'
+                % malformed_sender_email))
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True), email_exception:
+            gae_email_services.send_mail(
+                malformed_sender_email, self.RECIPIENT_EMAIL,
+                'subject', 'body', 'html')
+        messages = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAIL)
+        self.assertEqual(0, len(messages))
+
+    def test_that_email_not_sent_if_recipient_address_is_malformed(self):
+        malformed_recipient_email = ''
+        email_exception = (self.assertRaisesRegexp(
+            ValueError, 'Malformed recipient email address: %s'
+            % malformed_recipient_email))
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True), email_exception:
+            gae_email_services.send_mail(
+                feconf.ADMIN_EMAIL_ADDRESS, malformed_recipient_email,
+                'subject', 'body', 'html')
+
+
+class BulkEmailsTests(test_utils.GenericTestBase):
+    SENDER_EMAIL = 'sender@example.com'
+    SENDER_USERNAME = 'sender'
+    RECIPIENT_A_EMAIL = 'a@example.com'
+    RECIPIENT_A_USERNAME = 'usera'
+    RECIPIENT_B_EMAIL = 'b@example.com'
+    RECIPIENT_B_USERNAME = 'userb'
+    RECIPIENT_EMAILS = [RECIPIENT_A_EMAIL, RECIPIENT_B_EMAIL]
+
+    def setUp(self):
+        super(BulkEmailsTests, self).setUp()
+        # SENDER is authorised sender.
+        # A and B are recipients.
+        self.signup(self.SENDER_EMAIL, self.SENDER_USERNAME)
+        self.signup(self.RECIPIENT_A_EMAIL, self.RECIPIENT_A_USERNAME)
+        self.signup(self.RECIPIENT_B_EMAIL, self.RECIPIENT_B_USERNAME)
+        self.recipient_emails = [self.RECIPIENT_A_EMAIL, self.RECIPIENT_B_EMAIL]
+
+    def test_that_correct_bulk_emails_sent(self):
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_bulk_mail(
```

Comment:
@seanlip, I'm not sure whether we should directly call this function in testing? 

![image](https://user-images.githubusercontent.com/16653571/53747916-4a289400-3eca-11e9-8a7c-e057607f07b6.png)
(From the unit-test perspective I think it's fine but wanted to have a confirmation.)

Reply:
Oh yes, the comment does suggest against doing so. But it not being a private function, I assumed I can use it in writing Unit Test. If I am not supposed to do this, I believe we should change `send_bulk_mail` -> `_send_buk_mail` to make this clear for other developers.

Reply:
> we should change send_bulk_mail -> _send_buk_mail in the file gae_email_services.py to make this clear for other developers.

`send_bulk_mail` cannot be made as a private function as it's used in the email_manager.py, this is more like a developer should go through the docstring before using a method.  

Also, I'm still not sure whether we should allow using this function for testing? (ditto for gae_email_services.sent_email) @seanlip, any suggestion?

Reply:
It's fine to test it; it's part of the "public API" of this module.

Reply:
Thanks @seanlip!